cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+`cudf::string_view` is the data type of a `cudf::column` of type `STRING` just like `int32_t` is the data 
+type for a `cudf::column` of type `INT32`. As it's name implies, this is a read-only object instance
+that points to device memory inside the strings column. It's lifespan is the same (or less) as the 
+column it views.
+
+Use the `column_device_view::element` method to access an individual row element. Like any other
+column, do not call `element()` on a row that is null. 
+
+```c++
+   cudf::column_device_view d_strings;
+   ...
+   if( d_strings.is_valid(row_index) ) {
+      string_view d_str = d_strings.element<string_view>(row_index);
+      ...
+   }
+```
+
+A null string is not the same as an empty string. Use the `string_scalar` class if you need an 
+instance of a class object to represent a null string.
+
+The `string_view` contains comparison operators `<,>,==,<=,>=` that can be used in many cudf 
+functions like `sort` without string-specific code. The data for a `string_view` instance is 
+required to be [UTF-8](#UTF-8) and all operators and methods expect this encoding. Unless documented
+otherwise, position and length parameters are specified in characters and not bytes. The class also
+includes a `string_view::const_iterator` which can be used to navigate through individual characters
+within the string.
+
+`cudf::type_dispatcher` dispatches to the `string_view` data type when invoked on a `STRING` column.
+
+#### UTF-8
+
+The libcudf strings column only supports UTF-8 encoding for strings data. 
+[UTF-8](https://en.wikipedia.org/wiki/UTF-8) is a variable-length character encoding wherein each 
+character can be 1-4 bytes. This means the length of a string is not the same as its size in bytes.
+For this reason, it is recommended to use the `string_view` class to access these characters for
+most operations.
+
+The `string_view.cuh` header also includes some utility methods for reading and writing 
+(`to_char_utf8/from_char_utf8`) individual UTF-8 characters to/from byte arrays.
+
+### `cudf::lists_column_view` and `cudf::lists_view`
+
+`cudf::lists_column_view` is a view of a lists column. `cudf::list_view` is a view of a single list, and 
+therefore `cudf::list_view` is the data type of a `cudf::column` of type `LIST`.
+
+`cudf::type_dispatcher` dispatches to the `list_view` data type when invoked on a `LIST` column.
+
+### `cudf::structs_column_view` and `cudf::struct_view`
+
+`cudf::structs_column_view` is a view of a structs column. `cudf::struct_view` is a view of a single
+struct, and therefore `cudf::struct_view` is the data type of a `cudf::column` of type `STRUCT`.
+
+`cudf::type_dispatcher` dispatches to the `struct_view` data type when invoked on a `STRUCT` column.
+
+# cuIO: file reading and writing
+
+cuIO is a component of libcudf that provides GPU-accelerated reading and writing of data file 
+formats commonly used in data analytics, including CSV, Parquet, ORC, and Avro.
```

Comment:
and JSON Lines (not sure if omitted intentionally).

Reply:
Added